presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java

```diff
@@ -530,6 +525,8 @@ public PlanOptimizers(
                 costCalculator,
                 new TranslateExpressions(metadata, sqlParser).rules()));
 
+        builder.add(new ApplyConnectorOptimization(planOptimizerManager.getOptimizers()));
```

Comment:
It is not clear to me that placing connector-driven plan transformations at the very end is correct. How are you thinking about that?

- There is `// DO NOT add optimizers that change the plan shape (computations) after this point` a few lines above.
- It doesn't seem to make sense to add exchanges and partial/intermediate aggregations before the connector-based rules. I'm thinking of connector-based rules as transformation of the `logical` plan, not `physical` one. 
- It seems to make sense to run rules like `PruneUnreferencedOutputs` and `RemoveRedundantIdentityProjections` after connector-based rules. 
- What about DELETE queries? `MetadataDeleteOptimizer` and `BeginTableWrite` already ran.

Reply:
- Yes, ultimately, the rule will be moved all the way up to somewhere close to `AddExchange`. For now, we cannot achieve it because we need to gradually push `TranslateExpressions` up by migrating `Expression` to `RowExpression`.
- Given for now, we only focus on `FilterNode` and `TableScanNode` to support filter pushdown, it should be ok to have it after `HashGenerationOptimizer`? I will add to huge TODO to indicate the risk.
- We can move this above `MetadataDeleteOptimizer` and `BeginTableWrite` given there is no expression/rowexpression issue.